File: app/src/main/java/com/example/application/netwotk/BaseOkhttp.java
package com.example.application.netwotk;

import android.text.TextUtils;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class BaseOkhttp {
    private static BaseOkhttp mBaseOkhttp;
    private final OkHttpClient client = new OkHttpClient();

    private static final int X_RESPONSE_CODE = 200;

    private String mUrl;

    private BaseInterface mBaseInterface;

    public static BaseOkhttp getInstance() {
        if (mBaseOkhttp == null) {
            synchronized (BaseOkhttp.class) {
                if (mBaseOkhttp == null) {
                    mBaseOkhttp = new BaseOkhttp();
                }
            }
        }
        return mBaseOkhttp;
    }

    public void loadUrl(String url) {
        this.mUrl = url;
    }

    public void loadData() {
        load(null);
    }

    public void loadData(RequestBody requestBody) {
        load(requestBody);
    }


    private void load(RequestBody requestBody) {
        if (TextUtils.isEmpty(mUrl)) throw new NullPointerException("发送网络请求，请调用loadUrl函数加载URL");
        final Request.Builder builder = new Request.Builder();
        if (null == requestBody) {
            builder.get();
        } else {
            builder.post(requestBody);
        }
        Request request = builder.url(mUrl).build();
        Call call = client.newCall(request);
        call.enqueue(new Callback() {
            @Override
            public void onFailure(@NotNull Call call, @NotNull IOException e) {
                mBaseInterface.BaseErrorListener(e.getMessage());
            }

            @Override
            public void onResponse(@NotNull Call call, @NotNull Response response) throws IOException {
                int code = response.code();
                String data = response.body().string();
                if (code == X_RESPONSE_CODE) {
                    mBaseInterface.BaseSuccessListener(data);
//                    if (TextUtils.isEmpty(data)){
//                        mBaseInterface.BaseSuccessListener(data);
//                    }else {
//                        mBaseInterface.BaseErrorListener();
//                    }

                } else {
                    mBaseInterface.BaseErrorListener(data);
                }
            }
        });
    }

    public void setBaseInterfaceListener(BaseInterface baseInterfaceListener) {
        this.mBaseInterface = baseInterfaceListener;
    }
}

File: app/src/main/java/com/example/application/netwotk/BaseOkhttp.kt
package com.example.application.netwotk

import android.text.TextUtils
import android.util.Log
import com.example.application.netwotk.interfaces.BaseInterface
import okhttp3.*
import java.io.IOException

class BaseOkhttp {
    private var mUrl: String? = null
    private var mBaseInterface: BaseInterface? = null
    fun loadUrl(url: String?) {
        mUrl = url
    }

    fun loadData(map: Map<String, Any>?) {
        loadGet(map)
    }

    fun loadData(requestBody: RequestBody?) {
        loadPost(requestBody)
    }

    private fun loadGet(query: Map<String, Any>?) {
        if (TextUtils.isEmpty(mUrl)) throw NullPointerException("发送网络请求，请调用loadUrl函数加载URL")
        val builder = Request.Builder()
        val request = builder.url(mUrl!!).build()
        if (query != null) {
            val urlBuilder = request.url.newBuilder()
            for ((key, value) in query) {
                urlBuilder.addQueryParameter(key, value as String)
            }
        }
        val call = client.newCall(request)
        call.enqueue(callback)
    }

    private fun loadPost(requestBody: RequestBody?) {
        if (TextUtils.isEmpty(mUrl)) throw NullPointerException("发送网络请求，请调用loadUrl函数加载URL")
        val builder = Request.Builder()
        if (null != requestBody) {
            builder.post(requestBody)
        }
        val request = builder.url(mUrl!!).build()
        val call = client.newCall(request)
        call.enqueue(callback)
    }

    private val callback: Callback = object : Callback {
        override fun onFailure(call: Call, e: IOException) {
            mBaseInterface!!.BaseErrorListener(e.message)
        }

        @Throws(IOException::class)
        override fun onResponse(
            call: Call,
            response: Response
        ) {
            val data = response.body!!.string()
            Log.i("onResponse", data)
            if (response.isSuccessful) {
                mBaseInterface!!.BaseSuccessListener(data)
                //                    if (TextUtils.isEmpty(data)){
//                        mBaseInterface.BaseSuccessListener(data);
//                    }else {
//                        mBaseInterface.BaseErrorListener();
//                    }
            } else {
                val code = response.code
                mBaseInterface!!.BaseErrorListener(data)
            }
        }
    }

    fun setBaseInterfaceListener(baseInterfaceListener: BaseInterface?) {
        mBaseInterface = baseInterfaceListener
    }

    companion object {
        private var mBaseOkhttp: BaseOkhttp? = null
        private val client = OkHttpClient()
        private const val X_RESPONSE_CODE = 200
        @JvmStatic
        val instance: BaseOkhttp?
            get() {
                if (mBaseOkhttp == null) {
                    synchronized(BaseOkhttp::class.java) {
                        if (mBaseOkhttp == null) {
                            mBaseOkhttp = BaseOkhttp()
                        }
                    }
                }
                return mBaseOkhttp
            }
    }
}